tensorflow/core/kernels/scatter_op.h

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2015 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_KERNELS_SCATTER_OP_H_
+#define TENSORFLOW_KERNELS_SCATTER_OP_H_
+
+// Functor definitions for Scatter ops, must be compilable by nvcc.
+
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor_types.h"
+
+namespace tensorflow {
+
+class OpKernelContext;
+
+namespace scatter_op {
+
+enum class UpdateOp { ASSIGN, ADD, SUB };
+
+template <UpdateOp Op>
+struct Assign {};
+template <>
+struct Assign<UpdateOp::ASSIGN> {
+  template <typename Params, typename Update>
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
+  static void Run(Params&& p, const Update& u) {
+    p = u;
+  }
+};
+template <>
+struct Assign<UpdateOp::ADD> {
+  template <typename Params, typename Update>
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
+  static void Run(Params&& p, const Update& u) {
+    p += u;
```

Comment:
I used universal reference because CPU code does
```params.template chip<0>(indices(i)) = updates.template chip<0>(i)```
where the LHS is a rvalue.  If I used a pointer, I would have to first store the LHS in a temporary variable, then pass its address.  Besides the code looking ugly, it may also be less efficient.

It felt like universal reference is designed to solve exactly this problem.  But, there's a simple workaround, which is to store the LHS in a temporary variable, then use plain old references.  I did this now.  PTAL.

I will look into the thread-safety next.  Thanks for pointing to Thrust.  I also realized how to do range checks efficiently now without temporary overwriting data - just write a fast range check kernel.  I'll do that as well.